tarantool/msgpack_ext/types/datetime.py

```diff
@@ -0,0 +1,237 @@
+        # https://github.com/python/cpython/blob/1756ffd66a38755cd45de51316d66266ae30e132/Lib/datetime.py#L1591
+        return tzinfo.tzname(timestamp)
+    
+    raise ValueError(f'Unsupported timezone type {type(tzinfo)}')
+
+
+def is_abbrev_tz(tzname):
+    return tzname in tt_timezones.timezoneAbbrevInfo
+
+def assert_nonambiguous_tz(tzname, tt_tzinfo, error_class):
+    if (tt_tzinfo['category'] & tt_timezones.TZ_AMBIGUOUS) != 0:
+        raise error_class(f'Failed to create datetime with ambiguous timezone "{tzname}"')
+
+def get_python_tzinfo(tzindex):
+    if tzindex not in tt_timezones.indexToTimezone:
+        raise MsgpackError(f'Failed to create datetime with unknown tzindex {tzindex}')
+
+    tzname = tt_timezones.indexToTimezone[tzindex]
+
+    try:
+        tzinfo = pytz.timezone(tzname)
+    except pytz.exceptions.UnknownTimeZoneError:
+        tt_tzinfo = tt_timezones.timezoneAbbrevInfo[tzname]
+        assert_nonambiguous_tz(tzname, tt_tzinfo, MsgpackError)
+
+        tzinfo = datetime.timezone(datetime.timedelta(minutes=tt_tzinfo['offset']),
+                                   name=tzname)
+
+    return tzinfo
+
+def validate_python_timezone(timestamp):
+    tzname = get_python_tzname(timestamp)
+
+    if (tzname is not None):
+        if tzname not in tt_timezones.timezoneToIndex:
+            raise ValueError(f'Failed to create datetime with unknown timezone "{tzname}"')
+
+        if not is_abbrev_tz(tzname):
+            return
+
+        tt_tzinfo = tt_timezones.timezoneAbbrevInfo[tzname]
+        assert_nonambiguous_tz(tzname, tt_tzinfo, ValueError)
+
+def msgpack_decode(data):
+    cursor = 0
+    seconds, cursor = get_bytes_as_int(data, cursor, SECONDS_SIZE_BYTES)
+
+    if len(data) > SECONDS_SIZE_BYTES:
+        nsec, cursor     = get_bytes_as_int(data, cursor, NSEC_SIZE_BYTES)
+        tzoffset, cursor = get_bytes_as_int(data, cursor, TZOFFSET_SIZE_BYTES)
+        tzindex, cursor  = get_bytes_as_int(data, cursor, TZINDEX_SIZE_BYTES)
+    elif len(data) == SECONDS_SIZE_BYTES:
+        nsec = 0
+        tzoffset = 0
+        tzindex = 0
+    else:
+        raise MsgpackError('Unexpected datetime payload length')
+
+    total_nsec = seconds * NSEC_IN_SEC + nsec
+
+    if (tzindex != 0):
+        tzinfo = get_python_tzinfo(tzindex)
+        timestamp = pandas.to_datetime(total_nsec, unit='ns')\
+                          .replace(tzinfo=datetime.timezone.utc)\
+                          .tz_convert(tzinfo)
+    elif (tzoffset != 0):
+        tzinfo = datetime.timezone(datetime.timedelta(minutes=tzoffset))
+        return pandas.to_datetime(total_nsec, unit='ns')\
+                     .replace(tzinfo=datetime.timezone.utc)\
+                     .tz_convert(tzinfo)
+    else:
+        # return timezone-naive pandas.Timestamp
+        return pandas.to_datetime(total_nsec, unit='ns')
+
+    return timestamp
+
+class Datetime():
+    def __init__(self, *args, tarantool_tzindex=None, **kwargs):
```

Comment:
Do we need tarantool_tzindex?

Reply:
No, sorry, I forgot to clean it up. Fixed.